Restore call stack

gabi-250 opened the pull request:
This introduces the code necessary to restore the state of each frame whenever a guard fails. This should also restore the register state.

To implement this, it was also necessary to:
* remove the `MachineFunctionPass` I'd implemented and modify the `emitPrologue` and `emitEpilogue` code in `lib/Target/X86/X86FrameLowering.cpp` to ensure each optimized function has the same stack size as its unoptimized version
* implement a crude version of live variable analysis in `CheckPointPass.cpp` (closes #2)
* insert a `stackmap` call after each call to a function defined in the module being processed (also implemented in `CheckPointPass`). This allows the runtime to obtain a record of all the locations that are live at each call site. The address of the stackmap is automatically recorded, which allows the runtime to work out the return addresses of the 'unoptimized' functions. (closes #4)

I've done my best to document my code, but I am aware it may not always be obvious what it does. Please let me know which parts are unclear/should be reimplemented.

(Also closes #6 as far as I can tell)

vext01 commented on src/passes/function_passes/CheckPointPass.cpp:
```diff
@@ -12,179 +12,220 @@
   CheckPointPass() : FunctionPass(id) {}
 
+  virtual bool doInitialization(Module &mod) {
+    // Declare the guard failure function. This is necessary because the
```
So this is like an `extern` in C?

gabi-250 replied:
Yes, that is correct.

vext01 commented on src/passes/function_passes/CheckPointPass.cpp:
```diff
@@ -12,179 +12,220 @@
       for (BasicBlock::iterator it = bb.begin(); it != bb.end(); ++it) {
-        if (isa<CallInst>(it) &&
-            cast<CallInst>(*it).getCalledFunction()->getName() == "putchar") {
+        if (funName.endswith("more_indirection") && isa<ReturnInst>(it)) {
```
There's a hardcoded function name in here. We should at least put an `XXX`.

gabi-250 replied:
I can do that, but the point is that a 'guard' is to going inserted at that particular point in the code. There is no real logic behind why I've chosen that particular condition.

vext01 commented on src/passes/function_passes/CheckPointPass.cpp:
```diff
@@ -12,179 +12,220 @@
-            // which represents the number of arguments of the callback, and
-            // the arguments to pass to the callback.
+          if (!funName.startswith(UNOPT_PREFIX)) {
+            // The current function will optimized -> insert a patchpoint call.
```
`he current function will optimized`. Seems like a typo?

gabi-250 replied:
Well, this pass runs 'as early as possible', so at this point in time, both the `__unopt_` version and the regular versions are identical. So really, none of the functions are optimized at this time

vext01 commented on src/passes/function_passes/UnoptimizedCopyPass.cpp:
```diff
@@ -43,27 +44,33 @@ struct UnoptimizedCopyPass: public FunctionPass {
-              if (new_fun) {
-                call.setCalledFunction(new_fun);
+    auto funName = fun.getName();
+    if (funName == TRACE_FUN_NAME || funName == "more_indirection"
```
Similarly here, add an `XXX`?

gabi-250 replied:
Can do, but it is probably best to remove everything after `||`. The only function we really **don't** want inlined is `trace`. I disabled all inlining to test my code, that's why there are some hardcoded values here and there.

vext01 commented on src/stackmap_checker/guard.c:
```diff
@@ -0,0 +1,350 @@
+#define UNW_LOCAL_ONLY
```
Better add a license to new files.

vext01 replied:
What does `UNW_LOCAL_ONLY` do?

gabi-250 replied:
It is supposed to run an optimized version of `libunwind`:

```
However, if you tell libunwind that your program only needs local unwinding, then a special implementation can be selected which may run much faster than the generic implementation which supports both kinds of unwinding. To select this optimized version, simply define the macro UNW_LOCAL_ONLY before including the headerfile <libunwind.h>.
```

vext01 replied:
OK. I wonder what "non-local" unwinding would mean?

gabi-250 replied:
I think anything outside of the current compilation unit is 'non-local', but it's just a guess.

vext01 commented on src/stackmap_checker/guard.c:
```diff
@@ -0,0 +1,350 @@
+
+        if (!registers[frame]) {
+            // 16 registers are saved for each frame
+            registers[frame] = calloc(16, sizeof(unw_word_t));
```
Have you used valgrind on this code to check there are no memory errors?

gabi-250 replied:
Yes, I have.

vext01 replied:
Good job!

vext01 commented on src/stackmap_checker/guard.c:
```diff
@@ -0,0 +1,350 @@
+        unw_get_reg(&cursor, UNW_X86_64_R12, &registers[frame][12]);
+        unw_get_reg(&cursor, UNW_X86_64_R13, &registers[frame][13]);
+        unw_get_reg(&cursor, UNW_X86_64_R14, &registers[frame][14]);
+        unw_get_reg(&cursor, UNW_X86_64_R15, &registers[frame][15]);
```
This is pretty neat. Good find with libunwind.

vext01 commented on src/stackmap_checker/guard.c:
```diff
@@ -0,0 +1,350 @@
+        unw_get_reg(&cursor, UNW_X86_64_R15, &registers[frame][15]);
+        // Find address of the return address.
+        ret_addrs[frame] =
+            (uint64_t)((char *)registers[frame][UNW_X86_64_RBP] + 8);
```
Why `char *`?

gabi-250 replied:
Ah, you're right, it's unnecessary.

vext01 commented on src/stackmap_checker/guard.c:
```diff
@@ -0,0 +1,350 @@
+    }
+
+    // Get the end address of the function in which a guard failed.
+    void *end_addr = get_sym_end((void *)opt_size_rec->fun_addr, "trace");
```
Another hard-coded function name here?

gabi-250 commented on src/passes/function_passes/CheckPointPass.cpp:
```diff
@@ -12,179 +12,220 @@
- * This only runs on `trace` and `__unopt_trace`. All other functions are
- * ignored.
+ * The unoptimized version of each function contains the same number of
+ * `llvm.experimental.stackmap` calls as the original version. Knowing the
```
The 'original' version is the user-defined function. The other function pass creates an `__unopt_` version of each user-defined function, which I call the 'unoptimized' version.

gabi-250 commented:
@vext01 Thank you for your review! I think I've addressed all your comments, except for the one regarding the license. I haven't added a license to any of the files yet, but I will do before this PR is merged.

vext01 commented:
You can do licenses in a separate PR if you like.

Ready to squash?

gabi-250 commented:
@vext01 Sure. Is there nothing else you'd like changed? I am aware this is hard to review and that my code isn't that well written, but please **do** let me know if you see anything awfully wrong with it!

gabi-250 commented:
If not, this is squashed and ready to merge.

vext01 commented:
I'd be here all for days doing a deep review of these changes. I'm good to merge so you can continue.
